File: 152. Maximum Product Subarray/solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxValue=INT_MIN;
        int recordNum = 1;
            
        for(int i=0 ; i<nums.size() ; ++i){
            recordNum *= nums[i];
            maxValue = max(maxValue, recordNum);
            if(nums[i] == 0) recordNum=1;
        }
        
        recordNum = 1;
        for(int i=nums.size()-1 ; i>=0 ; --i){
            recordNum *= nums[i];
            maxValue = max(maxValue, recordNum);
            if(nums[i] == 0) recordNum=1;
        }
        
        return maxValue;
    }
};

File: 152. Maximum Product Subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_value = float("-inf")
        
        tmp_val = 1
        for num in nums:
            tmp_val *= num
            max_value = max(max_value, tmp_val)
            if num == 0:
                tmp_val = 1
            if num > max_value:
                max_value = num
        
        tmp_val = 1
        for num in nums[::-1]:
            tmp_val *= num
            max_value = max(max_value, tmp_val)
            if num == 0:
                tmp_val = 1
            if num > max_value:
                max_value = num
        
        return max_value
